test/MOI_callbacks.jl

```diff
@@ -0,0 +1,397 @@
+                MOI.LessThan(5.0)
+            )
+        end)
+        @test_throws(
+            ErrorException(
+                "`MOI.LazyConstraint` can only be called from LazyConstraintCallback."
+            ),
+            MOI.optimize!(model)
+        )
+    end
+    @testset "HeuristicSolution" begin
+        model, x, item_weights = callback_knapsack_model()
+        MOI.set(model, MOI.UserCutCallback(), cb_data -> begin
+            MOI.submit(
+                model,
+                MOI.HeuristicSolution(cb_data),
+                [x[1]],
+                [0.0]
+            )
+        end)
+        @test_throws(
+            ErrorException(
+                "`MOI.HeuristicSolution` can only be called from HeuristicCallback."
+            ),
+            MOI.optimize!(model)
+        )
+    end
+end
+
+@testset "HeuristicCallback" begin
+    @testset "HeuristicSolution" begin
+        model, x, item_weights = callback_knapsack_model()
+        solution_accepted = false
+        solution_rejected = false
+        MOI.set(model, MOI.HeuristicCallback(), cb_data -> begin
+            x_vals = MOI.get.(model, MOI.CallbackVariablePrimal(cb_data), x)
+            if MOI.submit(
+                model,
+                MOI.HeuristicSolution(cb_data),
+                x,
+                floor.(x_vals)
+            ) == MOI.HEURISTIC_SOLUTION_ACCEPTED
+                solution_accepted = true
+            end
+            if MOI.submit(
+                model,
+                MOI.HeuristicSolution(cb_data),
+                x,
+                ceil.(x_vals)
+            ) == MOI.HEURISTIC_SOLUTION_REJECTED
+                solution_rejected = true
+            end
+        end)
+        MOI.optimize!(model)
+        @test solution_accepted
+        @test solution_rejected
+    end
+    @testset "LazyConstraint" begin
+        model, x, item_weights = callback_knapsack_model()
+        MOI.set(model, MOI.HeuristicCallback(), cb_data -> begin
+            MOI.submit(
+                model,
+                MOI.LazyConstraint(cb_data),
+                MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(1.0, x), 0.0),
+                MOI.LessThan(5.0)
+            )
+        end)
+        @test_throws(
+            ErrorException(
+                "`MOI.LazyConstraint` can only be called from LazyConstraintCallback."
+            ),
+            MOI.optimize!(model)
+        )
+    end
+    @testset "UserCut" begin
+        model, x, item_weights = callback_knapsack_model()
+        MOI.set(model, MOI.HeuristicCallback(), cb_data -> begin
+            MOI.submit(
+                model,
+                MOI.UserCut(cb_data),
+                MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(1.0, x), 0.0),
+                MOI.LessThan(5.0)
+            )
+        end)
+        @test_throws(
+            ErrorException(
+                "`MOI.UserCut` can only be called from UserCutCallback."
```

Comment:
We should have an MOI error for that. It's likely to be thrown from every solver with a solver-independent error.

Reply:
We could even have a default fallback
```julia
# in MOI
MOI.submit(::MOI.ModelLike, ::UserCut{<:AbstractLazyConstraintCallbackData}, args...) = throw(...)
# in Gurobi
struct UserCutCallbackData ... end
MOI.submit(::Optimizer, ::UserCut{UserCutCallbackData}, ...) = ...
```

Reply:
:+1: on a MOI error.

:-1: on different callback data for each callback data, because that doesn't solve
```Julia
MOI.set(model, MOI.LazyConstraintCallback(), cb_data -> begin
    MOI.submit(model, MOI.UserCut{Gurobi.UserCutCallbackData}(cb_data), f, s)
end)
```

Reply:
The idea is that the type of `cb_data` will be `LazyConstraintCallbackData` so the user will have to do:
```julia
MOI.set(model, MOI.LazyConstraintCallback(), cb_data -> begin
    MOI.submit(model, MOI.UserCut(Gurobi.UserCutCallbackData(cb_data.inner_data))), f, s)
end)
```
With the current proposal, he can also do
```julia
MOI.set(model, MOI.LazyConstraintCallback(), cb_data -> begin
    model.callback_state = Gurobi.CB_USER_CUT
    MOI.submit(model, MOI.UserCut(cb_data), f, s)
end)
```

Reply:
> With the current proposal, he can also do

Sure. But I can do anything modifying internal fields. We still need the `model.callback_state` field to catch `OptimizeInProgress`, so the choice is between making this a `Bool` and having lots of different `XXXCallbackData` types, or making it an enum.

I'll stick with the enum for now.

Reply:
Let's keep it simple for now on this PR I agree. This thread is more at the stage of brainstorming future ideas as it won't have impact on user facing interface.